### PR TITLE
[HW] Address new HWModule builder review comments

### PR DESCRIPTION
**Pull request by mortbopet**

### CODE REVIEW

File: lib/Conversion/CalyxToHW/CalyxToHW.cpp [hunk 1]
```diff
@@ -42,40 +42,36 @@ struct ConvertComponentOp : public OpConversionPattern<ComponentOp> {
...
+              ports.setOutput(name, wireRead);
+              break;
+            }
+          }
+        });
+
+    auto *outputOp = hwMod.getBodyBlock()->getTerminator();
     rewriter.mergeBlocks(component.getBody(), hwMod.getBodyBlock(), argValues);
-    rewriter.create<OutputOp>(component.getLoc(), outputWires);
+    outputOp->moveAfter(&hwMod.getBodyBlock()->back());
```
Comment (teqdruid) [inline lib/Conversion/CalyxToHW/CalyxToHW.cpp:73]: Why is this necessary?

**Review by teqdruid — COMMENTED** [review]
lgtm! thanks for the test

File: lib/Conversion/CalyxToHW/CalyxToHW.cpp (hunk 1 above)
Comment (mikeurbach) [reply lib/Conversion/CalyxToHW/CalyxToHW.cpp:73]: I assume it's because the builder used to create the HWModuleOp now creates it automatically? I think this is generally unsafe, and the mutation here should be wrapped in `rewriter.updateRootInPlace(hwMod, ...)`

Comment (mikeurbach) [reply lib/Conversion/CalyxToHW/CalyxToHW.cpp:73]: Also, for the record, I don't think it's safe to use `b` in the body of the lambda that creates the HWModuleOp. I'm pretty sure all IR mutation needs to happen through `rewriter`

File: lib/Conversion/CalyxToHW/CalyxToHW.cpp [hunk 2]
```diff
@@ -42,40 +42,36 @@ struct ConvertComponentOp : public OpConversionPattern<ComponentOp> {
...
+          for (auto [name, type, direction, _] : portInfo) {
+            switch (direction) {
+            case calyx::Direction::Input:
+              assert(ports.getInput(name).getType() == type);
+              argValues.push_back(ports.getInput(name));
+              break;
+            case calyx::Direction::Output:
+              auto wire = b.create<sv::WireOp>(component.getLoc(), type, name);
+              auto wireRead =
+                  b.create<sv::ReadInOutOp>(component.getLoc(), wire);
```
Comment (richardxia) [inline lib/Conversion/CalyxToHW/CalyxToHW.cpp:63]: As I mentioned in https://github.com/llvm/circt/pull/3858, there were a few safety issues that @mikeurbach noticed when looking at this patch. One of them is these two lines which use the `OpBuilder` `b` to create wires. Paraphrasing what he told me, when using a pattern rewriter, you have to remember to use the rewriter to create new ops and not the underlying OpBuilder, since the rewriter needs to keep track of changes.

Locally, I tried changing both of these lines to `rewriter.create()`, but it didn't fix the crashes I was seeing. It sounds like we should still use `rewriter` over `b` here, but we may have other safety issues in this patch.

Comment (mortbopet) [reply lib/Conversion/CalyxToHW/CalyxToHW.cpp:63]: > One of them is these two lines which use the OpBuilder b to create wires. Paraphrasing what he told me, when using a pattern rewriter, you have to remember to use the rewriter to create new ops and not the underlying OpBuilder, since the rewriter needs to keep track of changes.

As far as i can tell, this is already what is being done. When one is using `create` with an `OpBuilder`-derived class, it is that same `OpBuilder` which is passed to the create method, see:
https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/IR/Builders.h#L456

Hence changing `b` to `rewriter` should have no effect.

Comment (mikeurbach) [reply lib/Conversion/CalyxToHW/CalyxToHW.cpp:63]: I wrote this on the other comment thread, but I think the concern with this specific case is with object slicing. By the time `create` calls the builder, it's just passing an `OpBuilder`: https://github.com/llvm/circt/blob/a610303a74af6d5fa48dc6c4810ab6b9b3ea0f2b/lib/Dialect/HW/HWOps.cpp#L601. That gets used inside the lambda, which means all the stuff on `ConversionPatternRewriter` gets sliced off inside the lambda, so the internal tracking that dialect conversion uses wouldn't work.

I'm not sure this is the root cause of the issue here, but in general it is not valid to mutate IR in a dialect conversion pass via any means other than a ConversionPatternRewriter. That's why in the other comment I also pointed out a situation to use `updateRootInPlace`.

Comment (dtzSiFive) [reply lib/Conversion/CalyxToHW/CalyxToHW.cpp:63]: Where does the slicing happen? (in the code you've posted it's passed by reference which doesn't cause slicing)

Comment (mikeurbach) [reply lib/Conversion/CalyxToHW/CalyxToHW.cpp:63]: Ok the slicing might be a red herring then, because I think builders/rewriters always get passed by reference. That would also explain why switching from `b` to `rewriter` didn't make a difference earlier.
